packages/gmail/src/Adaptor.js

```diff
@@ -0,0 +1,344 @@
+    const currentIds = [];
+    const previousIds = Array.isArray(state.processedIds)
+      ? state.processedIds
+      : [];
+
+    if (isTesting) {
+      console.log('previousIds', previousIds);
+    }
+
+    let nextPageToken = null;
+
+    do {
+      const messagesResult = await fetchMessages(userId, query, nextPageToken);
+      if (!messagesResult.messages?.length) {
+        console.log('No messages found.');
+        break;
+      }
+
+      nextPageToken = messagesResult.nextPageToken;
+
+      const incomingIds = messagesResult.messages.map(message => message.id);
+
+      const unprocessedIds = incomingIds.filter(
+        id => !previousIds.includes(id)
+      );
+
+      for (let messageId of unprocessedIds) {
+        const messageContent = {
+          messageId: messageId,
+        };
+
+        for (let hint of desiredContents) {
+          const desiredContent =
+            typeof hint === 'object' ? hint : { type: hint };
+
+          if (!desiredContent.type) {
+            if (desiredContent.archive) {
+              desiredContent.type = 'archive';
+            } else if (desiredContent.file) {
+              desiredContent.type = 'file';
+            }
+          }
+
+          if (!desiredContent.type) {
+            console.error('Unable to determine desired content type:', hint);
+            throw new Error('No desired content type provided.');
+          }
+
+          if (!desiredContent.name) {
+            desiredContent.name = desiredContent.type;
+          }
+
+          const content = await getContentFromMessage(
+            userId,
+            messageId,
+            desiredContent
+          );
+
+          if (content && messageContent[desiredContent.name]) {
+            throw new Error(
+              `Duplicate content name detected: ${desiredContent.name}`
+            );
+          }
+          
+          messageContent[desiredContent.name] ??= content;
+        }
+
+        messageContents.push(messageContent);
+      }
+
+      currentIds.push(...incomingIds);
+    } while (nextPageToken);
+
+    const expiredIds = previousIds.filter(id => !currentIds.includes(id));
+    const newIds = currentIds.filter(id => !expiredIds.includes(id));
+
+    const nextState = {
+      ...composeNextState(state, messageContents),
+      processedIds: newIds,
+    };
+
+    return callback(nextState);
+  };
+}
+
+async function fetchMessages(userId, query, lastPageToken) {
+  let messagesResponse = null;
+
+  try {
+    messagesResponse = await gmail.users.messages.list({
+      userId: userId,
+      q: query,
+      maxResults: 3,
+      pageToken: lastPageToken,
+    });
+  } catch (error) {
+    throw new Error('Error fetching messages: ' + error.message);
+  }
+
+  const { messages, nextPageToken } = messagesResponse.data;
+
+  return { messages, nextPageToken };
+}
+
+async function getContentFromMessage(userId, messageId, desiredContent) {
+  const messageResponse = await gmail.users.messages.get({
```

Comment:
The thing that concerns me here is that this function gets called once for every desiredContent type. So if you're passing an array of 8 fields, then each email will be downloaded 8 times.

I don't know if the gmail package caches the emails it downloads, but I doubt it.

I think this is important and could have a big impact on performance. I think your options are:
1) Pass all desiredContents in as an array, and use `Array.find()` or `Array.includes` to find all the content types you need. Then this only gets called once.
2) Add a caching layer so that messages are cached by id in-memory in the adaptor, and you only call out to gmail if the message isn't in the cache. This is probably the quickest option for you.